server/src/main/java/io/crate/role/TransportPrivilegesAction.java

```diff
@@ -91,61 +92,76 @@ protected void masterOperation(PrivilegesRequest request,
-                    if (unknownUserNames.isEmpty()) {
-                        affectedRows = applyPrivileges(mdBuilder, request);
+                new AckedClusterStateUpdateTask<>(Priority.IMMEDIATE, request, listener) {
+
+                    long affectedRows = -1;
+                    List<String> uknownRoleNames = null;
+
+                    @Override
+                    public ClusterState execute(ClusterState currentState) throws Exception {
+                        Metadata currentMetadata = currentState.metadata();
+                        Metadata.Builder mdBuilder = Metadata.builder(currentMetadata);
+                        uknownRoleNames = validateRoleNames(currentMetadata, request);
+                        if (uknownRoleNames.isEmpty()) {
+                            affectedRows = applyPrivileges(mdBuilder, request);
+                        }
+                        return ClusterState.builder(currentState).metadata(mdBuilder).build();
                     }
-                    return ClusterState.builder(currentState).metadata(mdBuilder).build();
-                }
 
-                @Override
-                protected PrivilegesResponse newResponse(boolean acknowledged) {
-                    return new PrivilegesResponse(acknowledged, affectedRows, unknownUserNames);
-                }
-            });
+                    @Override
+                    protected PrivilegesResponse newResponse(boolean acknowledged) {
+                        return new PrivilegesResponse(acknowledged, affectedRows, uknownRoleNames);
+                    }
+                });
 
     }
 
     @VisibleForTesting
-    static List<String> validateUserNames(Metadata metadata, Collection<String> userNames) {
+    static List<String> validateRoleNames(Metadata metadata, PrivilegesRequest request) {
         RolesMetadata rolesMetadata = metadata.custom(RolesMetadata.TYPE);
+        Set<String> roleNamesToCheck = new LinkedHashSet<>(request.userNames());
+        if (request.rolePrivilege() != null) {
+            roleNamesToCheck.addAll(request.rolePrivilege().roleNames());
+        }
         if (rolesMetadata == null) {
-            return new ArrayList<>(userNames);
+            return new ArrayList<>(roleNamesToCheck);
         }
-        List<String> unknownUserNames = null;
-        for (String userName : userNames) {
+        List<String> unknownRoleNames = null;
+        for (String roleName : roleNamesToCheck) {
             //noinspection PointlessBooleanExpression
-            if (rolesMetadata.roleNames().contains(userName) == false) {
-                if (unknownUserNames == null) {
-                    unknownUserNames = new ArrayList<>();
+            if (rolesMetadata.roleNames().contains(roleName) == false) {
+                if (unknownRoleNames == null) {
+                    unknownRoleNames = new ArrayList<>();
                 }
```

Comment:
Can we make unknownRoleNames an empty list from the beginning?  
Look like we never return null anyway.

Reply:
of course, thx.